src/main/java/org/zalando/nakadi/service/CursorOperationsService.java

```diff
@@ -0,0 +1,209 @@
+
+            // assume all timelines have an equal number of partitions
+            for (int i = 0; i < oldestStats.size(); i++) {
+                final PartitionStatistics oldStat = oldestStats.get(i);
+                final PartitionStatistics newStat = newestStats.get(i);
+                final NakadiCursorLag nakadiCursorLag = new NakadiCursorLag(oldStat.getTopic(), oldStat.getPartition(),
+                        oldStat.getFirst(), newStat.getLast());
+                stats.put(oldStat.getPartition(), nakadiCursorLag);
+            }
+
+            // assume all partitions are present in the `stats` map
+            return cursors.stream().map(cursor -> {
+                final NakadiCursorLag nakadiCursorLag = stats.get(cursor.getPartition());
+                final NakadiCursorDistanceQuery query = new NakadiCursorDistanceQuery(cursor,
+                        nakadiCursorLag.getLastCursor());
+                final NakadiCursorDistanceResult distance = this.calculateDistance(query);
+                nakadiCursorLag.setLag(distance.getDistance());
+                return nakadiCursorLag;
+            }).collect(Collectors.toList());
+        } catch (final NakadiException e) {
+            throw new MyNakadiRuntimeException1("error", e);
+        }
+    }
+
+    private List<PartitionStatistics> getStatsForTimeline(final Timeline timeline) throws ServiceUnavailableException {
+        return timelineService.getTopicRepository(timeline)
+                .loadTopicStatistics(Collections.singletonList(timeline));
+    }
+
+    public List<NakadiCursor> unshiftCursors(final List<ShiftedNakadiCursor> cursors) throws InvalidCursorOperation {
+        return cursors.stream().map(this::unshiftCursor).collect(Collectors.toList());
+    }
+
+    public NakadiCursor unshiftCursor(final ShiftedNakadiCursor cursor) throws InvalidCursorOperation {
+        if (cursor.getShift() < 0) {
+            return moveCursorBackwards(cursor.getTopic(), cursor.getTimeline(), cursor.getPartition(),
+                    numberOfEventsBeforeCursor(cursor),
+                    cursor.getShift());
+        } else {
+            return moveCursorForward(cursor.getTopic(), cursor.getTimeline(), cursor.getPartition(),
+                    numberOfEventsBeforeCursor(cursor), cursor.getShift());
+        }
+    }
+
+    private NakadiCursor moveCursorBackwards(final String topic, final Timeline timeline, final String partition,
+                                             final long offset, final long shift) {
+        final long shiftedOffset = offset + shift;
+        if (shiftedOffset >= 0) { // move left in the same timeline
+            final String paddedOffset = getOffsetForPosition(timeline, shiftedOffset);
+            return new NakadiCursor(timeline, partition, paddedOffset);
+        } else { // move to previous timeline
+            final Timeline previousTimeline = getTimeline(topic, timeline.getOrder() - 1);
```

Comment:
I don't feel comfortable with calling getActiveTimelines() several times without need for that